File: rearrange_array_elements_by_sign.go
func rearrangeArray(nums []int) []int {
    ans := make([]int, len(nums))
    positive, negative := 0, 1

    for index := 0; index < len(nums); index++ {
        if nums[index] > 0 {
            ans[positive] = nums[index]
            positive += 2
        } else {
            ans[negative] = nums[index]
            negative += 2
        }
    }
    return ans
}

// Time Complexity : O(n)
// Space Complexity : O(n)

File: rearrange_array_elements_by_sign.cpp
class Solution {
public:
    vector<int> rearrangeArray(vector<int>& nums) {
        vector<int> ans(nums.size(), 0);
        int index = 0, positive = 0, negative = 1;

        while (index < nums.size()) {
            if (nums[index++] > 0) {
                ans[positive] = nums[index-1];
                positive += 2;
            } else {
                ans[negative] = nums[index-1];
                negative += 2;
            }
        }
        return ans;
    }
};

// Time Complexity : O(n)
// Space Complexity : O(n)